languages/js/src/helpers.ts

```diff
@@ -359,12 +324,63 @@ export function printError(e: Error) {
  *
  * @internal
  */
-export function isConstructor(f: unknown): boolean {
+export function isConstructor(f: unknown): f is Class {
   try {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
     // @ts-ignore
     Reflect.construct(String, [], f);
     return true;
   } catch (e) {
     return false;
   }
 }
+
+/**
+ * Type guard to test if a value is a [[`obj`]].
+ *
+ * @internal
+ */
+export function isObj(x: unknown): x is obj {
+  return typeof x === 'object' && x !== null;
+}
+
+/**
+ * Type guard to test if a value is an ES6 Map with string keys and PolarTerm
+ * values.
+ *
+ * @internal
+ */
+function isMapOfPolarTerms(x: unknown): x is Map<string, PolarTerm> {
+  if (!(x instanceof Map)) return false;
+  for (const [key, value] of x.entries()) {
+    if (typeof key !== 'string' || !isPolarTerm(value)) return false;
+  }
+  return true;
+}
+
+/**
+ * Type guard to test if a value is an Array of PolarTerms.
+ *
+ * @internal
+ */
+function isArrayOfPolarTerms(x: unknown): x is Array<PolarTerm> {
+  return !Array.isArray(x) || x.some((v: unknown) => !isPolarTerm(v));
```

Comment:
is this logic right? `isArrayOfPolarTerms("foo") = !Array.isArray("foo") || ... = !false || ... = true` ??

anyway it'd be nice to build these compound checks out of combinators
```javascript
const negate<A> = (pred: Pred<A>) => (a: A) => !pred(a);
const conjoin<A> = (p0: Pred<A>, p1: Pred<A>) => (a: A) => p0(a) && p1(a);
const disjoin<A> = (p0: Pred<A>, p1: Pred<A>) => (a: A) => p0(a) || p1(a);
const some<A> = (p: Pred<A>) => (x: A[]) => x.some(p);
const every<A> = (p: Pred<A>) => (x: A[]) => x.every(p);

const isArrayOfPolarTerms = conjoin(Array.isArray, every(isPolarTerm));
```

Reply:
This would be cool; did a very novice version of this in c40377d0455c3daf4a161bf5e6cc8edd9e0977f8. Gonna punt for now on putting any more work into cleaning up the checks — a lot of (all of?) them are validating the structure of code emitted by the VM, so we control both ends of the pipe and could _probably_ delete most of this validation and not lose too much sleep. The validation checks were very helpful when I was building the library and figuring out what the shape of the data was after it'd gone through the wasm vortex, but we could probably just assert the shape of the data and move on.